File: 2022/day14/python/solution.py
#!/usr/bin/env python3
import os

def parse_paths(text):
    """Parse rock paths and return set of rock positions."""
    rocks = set()
    for line in text.strip().split('\n'):
        points = line.split(' -> ')
        for i in range(len(points) - 1):
            x1, y1 = map(int, points[i].split(','))
            x2, y2 = map(int, points[i + 1].split(','))

            # Draw line from (x1, y1) to (x2, y2)
            if x1 == x2:  # vertical line
                for y in range(min(y1, y2), max(y1, y2) + 1):
                    rocks.add((x1, y))
            else:  # horizontal line
                for x in range(min(x1, x2), max(x1, x2) + 1):
                    rocks.add((x, y1))
    return rocks

def simulate_sand(blocked, max_y, floor=False):
    """
    Simulate one unit of sand falling.
    Returns the resting position, or None if sand falls into abyss.
    """
    x, y = 500, 0

    while True:
        # Check if sand has fallen below all rocks (into abyss)
        if not floor and y > max_y:
            return None

        # Try to move down
        if floor and y + 1 == max_y + 2:
            # Hit the floor
            return (x, y)
        elif (x, y + 1) not in blocked:
            y += 1
        # Try to move down-left
        elif (x - 1, y + 1) not in blocked:
            x -= 1
            y += 1
        # Try to move down-right
        elif (x + 1, y + 1) not in blocked:
            x += 1
            y += 1
        # Sand comes to rest
        else:
            return (x, y)

def part1(text):
    """Count sand units that come to rest before sand falls into abyss."""
    rocks = parse_paths(text)
    max_y = max(y for x, y in rocks)
    blocked = rocks.copy()
    count = 0

    while True:
        pos = simulate_sand(blocked, max_y)
        if pos is None:
            break
        blocked.add(pos)
        count += 1

    return count

def part2(text):
    """Count sand units until source is blocked (with floor)."""
    rocks = parse_paths(text)
    max_y = max(y for x, y in rocks)
    blocked = rocks.copy()
    count = 0

    while True:
        pos = simulate_sand(blocked, max_y, floor=True)
        blocked.add(pos)
        count += 1
        if pos == (500, 0):
            break

    return count

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    input_file = os.path.join(script_dir, '..', 'input.txt')

    with open(input_file) as f:
        text = f.read()

    print('Part 1:', part1(text))
    print('Part 2:', part2(text))

if __name__ == '__main__':
    main()

File: 2022/day14/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use List::Util qw(min max);

# Parse rock paths and return hash of blocked positions
sub parse_paths {
    my ($text) = @_;
    my %rocks;

    for my $line (split /\n/, $text) {
        next unless $line =~ /\S/;
        my @points = split / -> /, $line;

        for my $i (0 .. $#points - 1) {
            my ($x1, $y1) = split /,/, $points[$i];
            my ($x2, $y2) = split /,/, $points[$i + 1];

            if ($x1 == $x2) {
                # Vertical line
                for my $y (min($y1, $y2) .. max($y1, $y2)) {
                    $rocks{"$x1,$y"} = 1;
                }
            } else {
                # Horizontal line
                for my $x (min($x1, $x2) .. max($x1, $x2)) {
                    $rocks{"$x,$y1"} = 1;
                }
            }
        }
    }

    return \%rocks;
}

# Simulate one unit of sand falling
# Returns the resting position as "x,y" or undef if sand falls into abyss
sub simulate_sand {
    my ($blocked, $max_y, $floor) = @_;
    my ($x, $y) = (500, 0);

    while (1) {
        # Check if sand has fallen below all rocks (into abyss)
        if (!$floor && $y > $max_y) {
            return undef;
        }

        # Try to move down
        if ($floor && $y + 1 == $max_y + 2) {
            # Hit the floor
            return "$x,$y";
        } elsif (!exists $blocked->{"$x," . ($y + 1)}) {
            $y++;
        }
        # Try to move down-left
        elsif (!exists $blocked->{($x - 1) . "," . ($y + 1)}) {
            $x--;
            $y++;
        }
        # Try to move down-right
        elsif (!exists $blocked->{($x + 1) . "," . ($y + 1)}) {
            $x++;
            $y++;
        }
        # Sand comes to rest
        else {
            return "$x,$y";
        }
    }
}

# Count sand units that come to rest before sand falls into abyss
sub part1 {
    my ($text) = @_;
    my $rocks = parse_paths($text);

    # Find max y coordinate
    my $max_y = 0;
    for my $key (keys %$rocks) {
        my ($x, $y) = split /,/, $key;
        $max_y = $y if $y > $max_y;
    }

    my %blocked = %$rocks;
    my $count = 0;

    while (1) {
        my $pos = simulate_sand(\%blocked, $max_y, 0);
        last unless defined $pos;
        $blocked{$pos} = 1;
        $count++;
    }

    return $count;
}

# Count sand units until source is blocked (with floor)
sub part2 {
    my ($text) = @_;
    my $rocks = parse_paths($text);

    # Find max y coordinate
    my $max_y = 0;
    for my $key (keys %$rocks) {
        my ($x, $y) = split /,/, $key;
        $max_y = $y if $y > $max_y;
    }

    my %blocked = %$rocks;
    my $count = 0;

    while (1) {
        my $pos = simulate_sand(\%blocked, $max_y, 1);
        $blocked{$pos} = 1;
        $count++;
        last if $pos eq "500,0";
    }

    return $count;
}

sub main {
    my $script_dir = dirname(__FILE__);
    my $input_file = "$script_dir/../input.txt";

    open my $fh, '<', $input_file or die "Cannot open $input_file: $!";
    my $text = do { local $/; <$fh> };
    close $fh;

    print "Part 1: ", part1($text), "\n";
    print "Part 2: ", part2($text), "\n";
}

main();
